tests/php/src/Internal/DependencyManagement/AbstractServiceProviderTests.php

```diff
@@ -0,0 +1,128 @@
+use Automattic\WooCommerce\Internal\DependencyManagement\ExtendedContainer;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithConstructorArgumentWithoutTypeHint;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithDependencies;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithPrivateConstructor;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\DependencyClass;
+use League\Container\Definition\DefinitionInterface;
+
+/**
+ * Tests for AbstractServiceProvider.
+ */
+class AbstractServiceProviderTests extends \WC_Unit_Test_Case {
+
+	/**
+	 * The system under test.
+	 *
+	 * @var AbstractServiceProvider
+	 */
+	private $sut;
+
+	/**
+	 * The container used for tests.
+	 *
+	 * @var ExtendedContainer
+	 */
+	private $container;
+
+	/**
+	 * Runs before each test.
+	 */
+	public function setUp() {
+		$this->container = new ExtendedContainer();
+
+		$this->sut = new class() extends AbstractServiceProvider {
+			// phpcs:disable
+
+			/**
+			 * Public version of add_with_auto_arguments, which is usually protected.
+			 */
+			public function add_with_auto_arguments( string $class_name, $concrete = null, bool $shared = false ) : DefinitionInterface {
+				return parent::add_with_auto_arguments( $class_name, $concrete, $shared );
+			}
+
+			/**
+			 * The mandatory 'register' method (defined in the base class as abstract).
+			 * Not implemented because this class is tested on its own, not as a service provider actually registered on a container.
+			 */
+			public function register() {}
+
+			// phpcs:enable
+		};
+
+		$this->sut->setContainer( $this->container );
+	}
+
+	/**
+	 * @testdox 'add_with_auto_arguments' should throw an exception if an invalid class name is passed.
+	 *
+	 * @throws \Exception Invalid class name passed.
+	 */
+	public function test_add_with_auto_arguments_throws_on_non_class_passed() {
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage( "AbstractServiceProvider::addWithAutoArguments: error when reflecting class 'foobar': Class foobar does not exist" );
+
+		$this->sut->add_with_auto_arguments( 'foobar' );
+	}
+
+	/**
+	 * @testdox 'add_with_auto_arguments' should throw an exception if the passed class has a private constructor.
+	 *
+	 * @throws \Exception The passed class has a private constructor.
+	 */
+	public function test_add_with_auto_arguments_throws_on_private_constructor() {
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage( "AbstractServiceProvider::addWithAutoArguments: constructor of class '" . ClassWithPrivateConstructor::class . "' isn't public, instances can't be created." );
+
+		$this->sut->add_with_auto_arguments( ClassWithPrivateConstructor::class );
+	}
+
+	/**
+	 * @testdox 'add_with_auto_arguments' should throw an exception if the passed class has a constructor argument without type hint.
```

Comment:
Given that we don't use this annotation anywhere else I don't know that we should start now?

Reply:
Too late, I have used this in other PRs too...

Also I don't think it's a bad thing to use it, we would have to put a descriptive comment for the test anyway so if there's a dedicate annotation for that purpose, why not using it?

Reply:
While I do feel like it's a bit redundant I can't argue with your mastery of annotations throughout the tests you've written. In fact, I've even learned a thing or two about PHPUnit 😄 If this is something we want to do as a team perhaps we need to add a PHPCS rule for it too?